sklearn/model_selection/_validation.py

```diff
@@ -727,12 +727,23 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
     predictions = func(X_test)
     if method in ['decision_function', 'predict_proba', 'predict_log_proba']:
         n_classes = len(set(y))
-        predictions_ = np.zeros((_num_samples(X_test), n_classes))
-        if method == 'decision_function' and len(estimator.classes_) == 2:
-            predictions_[:, estimator.classes_[-1]] = predictions
-        else:
-            predictions_[:, estimator.classes_] = predictions
-        predictions = predictions_
+        if n_classes != len(estimator.classes_):
+            if len(predictions.shape) == 2 and \
+                    predictions.shape[1] != len(estimator.classes_):
+                raise ValueError('Shape of output predictions does not '
+                                 'match number of classes in fold. '
+                                 'Cannot reconcile different number of '
+                                 'classes in different folds. To fix this, '
+                                 'use a cross-validation technique resulting '
+                                 'in properly stratified folds.')
+            predictions_ = np.zeros((_num_samples(X_test), n_classes))
+            if method == 'decision_function' and len(estimator.classes_) <= 2:
+                predictions_[:, estimator.classes_[-1]] = predictions
+            else:
+                predictions_[:, estimator.classes_] = predictions
+            if method == 'decision_function' and n_classes == 2:
+                predictions_ = predictions_[:, 1]
```

Comment:
this line does not appear to be covered by tests.

Reply:
I'm a little bit confused by this if else sequence. Why do we need the first if in addition to the second if? doesn't that cover all cases?

Reply:
Let's continue the conversation here. Sorry about that.

> Is there a case with n_classes == 2 where len(estimator.classes_) > 2? It looks to me like these two ifs are always either both true or neither but I might be missing something.

No but there are cases when `n_classes != 2` and `len(estimator.classes_) <= 2`